### DISCUSSION TITLE
v4.4.0-rc1 (beta1)

### DISCUSSION THREAD

**Discussion by pit-ray** [pit-ray/win-vind#76] [News]
#### New
- Add the following word-motion which behave almost exactly like Vim. (#57, #75)
  
  |**ID**|**Feature**|**Emulation**|
  |:---:|:---|:---:|
  |**move_fwd_word**|words forward for normal mode.|`w`|
  |**move_fwd_word_simple**|words forward for visual mode.|`w`|
  |**move_bck_word**|words backward for normal mode.|`b`|
  |**move_bck_word_simple**|words backward for visual mode.|`b`|
  |**move_fwd_bigword**|WORDS forward.|`W`|
  |**move_bck_bigword**|WORDS backward.|`B`|
  |**move_end_word**|Forward to the end of words.|`e`|
  |**move_end_bigword**|Forward to the end of WORDS.|`E`|
  |**move_bckend_word**|Backward to the end of words.|`ge`|
  |**move_bckend_bigword**|Backward to the end of WORDS.|`gE`|

  These functions do not work in visual mode except for `w` and `b`, because they copy the text once and retrieve the text via the clipboard. `iskeyword` option is fixed to the default value of Vim in Windows and cannot change it currently.  
  There is an option `charbreak` to set the criteria for considering a Unicode character as a single character. 

  |ID|Type|Default|Note|
  |:---:|:---:|:---:|:---|
  |`charbreak`|str|grapheme|Mode for how to split a single Unicode character. The `grapheme` mode treats a combination character as a single character. The `codepoint` mode processes the combination character for each codepoint.|

#### Note
- If you want to use word motion (e.g. `w`, `B`, `e`) in MS Office Word, it is recommended to disable `Use smart paragraph selection`.  

  <img src="https://github.com/pit-ray/win-vind/blob/master/docs/imgs/opt_in_ms_word.png?raw=true" title="Option in MS Word" width=600>

<hr /><em>This discussion was created from the release <a href='https://github.com/pit-ray/win-vind/releases/tag/v4.4.0-rc1'>v4.4.0-rc1</a>.</em>

**Comment by pit-ray** [comment]
This is not an official 4.4.0 release, but please give us feedback for impressions or feelings.

**Comment by gilbh** [comment]
Hi,

I tried to install 4.4.0rc, and have two questions:
1) could it be that the version is still shown as 4.3.3? That is what I am getting in the About window, which is confusing to me.
2) what is the difference between [setup_win-vind_4.4.0-rc1_64bit.exe](https://github.com/pit-ray/win-vind/releases/download/v4.4.0-rc1/setup_win-vind_4.4.0-rc1_64bit.exe) and [win-vind_4.4.0-rc1_64bit_installer.zip](https://github.com/pit-ray/win-vind/releases/download/v4.4.0-rc1/win-vind_4.4.0-rc1_64bit_installer.zip)?

Thanks,
Gil

**Reply by pit-ray** [reply]
1. The version is a forget to update. It is not an official release, so please forgive me.
2. When we download a file with an `.exe` extension in Windows, receive multiple warnings in general browsers, such as Edge, Chorome, Firefox. Therefore, I wrapped `.exe` file with zip. To avoid warnings with the `.exe` extension, we must pay $xx per year and issue a code signing certificate. However, this will be difficult because there is no profit and no sponser in win-vind.

Thanks.

**Reply by gilbh** [reply]
I don't think there is any reason to spend money for seamless `.exe`, but maybe you should consider removing the `.exe.`, because to me at least it was confusing, and, as I understand from what you are writing, there's nothing good about it (compared with `.zip`)

**Reply by pit-ray** [reply]
I didn't mention it above, but the `.exe` is used in the installation of winget. And, it is exactly the same thing as the wrapped one with zip, so there should be no problem if you accidentally download it.

**Reply by gilbh** [reply]
> 1. The version is a forget to update. It is not an official release, so please forgive me.

Just to clarify: in itself it is not a big deal, it is just that I need to be able to discern between `4.3` and `4.4` when I am testing stuff. If it is not complicated, I think it is worth to trouble to update the version in `4.4` b/c easier to figure out which version I am working with. Thanks!!

**Reply by pit-ray** [reply]
Thanks for letting me know about every little thing.
To update the display version, only rewrite [CMakeLists.txt](https://github.com/pit-ray/win-vind/blob/master/CMakeLists.txt#L2). However, I will do in beta2.

**Comment by gilbh** [comment]
From what I am seeing right now, caret move is very slow to respond. Each `w` operation takes time, I can literally see the selection of the sentence goes on and then off, before the caret jumps forward. When I switch back to an older version of `win-vind` (4.2.1), the caret moves fast again, as it should.

This is not particular to MS Word, it happened also in a regular Notepad.

**Reply by pit-ray** [reply]
The newly implemented word-motion captures text through the global clipboard like `~` command and parses the text for movement. The moving behavior is the same as Vim because it uses the same algorithm for text parsing.
As you say, its performance is not good. However, I could not implement the `e` command without parsing.

In only the `w` and `b` commands, the same function as before are available with `move_fwd_word_simple`, `move_bck_word_simple`.

**Example**
Using light behavior `w` and `b` instead.
```vim
ennoremap w move_fwd_word_simple
ennoremap b move_bck_word_simple
```

Thanks.

**Reply by gilbh** [reply]
Thanks. When I added the two lines to the `.vindrc`, the movement of `b` and `w` improved. But this got me thinking: why not routing them by default to `move_bck_word_simple` and `move_fwd_word_simple`? What is good in their slowness? (In contrast, I understand that `e` cannot be faster...).

Best
Gil

**Reply by pit-ray** [reply]
In terms of guaranteeing that it works exactly the same as the original Vim, I did so to unify the `e`, `gE`, etc.

But as you say, I would consider making `w` and `b` exceptions to the default.

**Comment by gilbh** [comment]
I see. I suggest checking all other actions, such as `W` and `B`, that do not require global clipboard (I think?), to try and minimize action time.

I am also wondering if there are any optimizations or other workarounds that you think could improve performance. It is really slow now (but works well!).

Just a thought. I get it that the limitations here are from the system and not your code.

Thanks for everything.

Best, Gil

**Reply by pit-ray** [reply]
WORD-motion, such as `gE`, `W` or `B`, needs text parsing as well. However the fast version could not be implemented because Windows does not  provide white-space based word motion.

If there is a way to get the text around the caret without global clipboard (e.g. UI Automation, machine learning based character recognition), they could be faster. And the waiting time for the clipboard could be optimized a bit more.

It is still a beta version, so I will consider more.

Thanks

**Comment by gilbh** [comment]
Fantastic!
